Updates to Metadata Entry Section (#4)
PecanProject/betydbdoc-dataentry: adding_metadata_sql.md

## SQL

SQL can also be used for bulk uploads. However, SQL should primarily be used in the following cases:

* fix data errors 

> * remove duplicate records

Reviewer: Don't use "record" to refer to an SQL database row.  "Record" should only be used in the colloquial sense as in "We have a record of the sorghum yield for 2016 stored in a row of the yields table in the BETYdb database." or in the technical sense used in the PostgreSQL documentation (see the table at https://www.postgresql.org/docs/9.4/datatype-pseudo.html or https://www.postgresql.org/docs/9.4/plpgsql-declarations.html#PLPGSQL-DECLARATION-RECORDS).  

([Some database people feel especially strongly about this.](https://i.stack.imgur.com/oIoG7.jpg))

But more to the point, this list is redundant and vague: The second and third item are subsumed under the first, but "fix data errors" is too broad.  There are all sorts of data errors that can be fixed just as easily and more safely by using the BETYdb Rails app.  The opening statements of this section might better read "SQL should be used only when the data manipulation to be done either can't be done using the web interface or BETYdb-YABA or where doing so would be exceedingly tedious."  Then give examples, which might include
1. correcting systemic data errors involving large numbers of rows
2. assigning non-point geometries to sites [this might be possible in YABA, I don't know]
3. associating experiments with sites and treatments
4. associating sites with cultivars

It's probably worth noting which of the steps below can't be done in the web interface (6, 7, and 9, and assigning a non-point geometry in 2) or the YABA interface.

> * remove duplicate records

Reviewer 2: ```suggestion
  * remove duplicates
  * transform values
```

* replace records
* metadata upload requires features that are not available through the web interface or BETYdb-YABA

### Metadata Entry Workflow for BETYdb

The following steps have been implemented in the BETYdb-YABA API and python client, but are described below for reference

Here we show SQL code used to add metadata required each season by the TERRA REF database, including experiments, sites, treatments, cultivars, and citations that is required prior to uploading the trait or yield data. Most PEcAn users _will not_ need to add experiments and cultivars, or associate these records with sites.

Also note that in the TERRA REF and other agronomic applications, each record in the 'sites' table will correspond to an experimental plot.

> 

Reviewer: All the SQL statements are missing semicolons, though depending on the context, a command can also be terminated by the end of the input stream.  If these commands are meant to be tried out in an interactive psql session, however, the semicolons are needed.

Reviewer 2: fixed

### Step 1: Add new experiments

> 

Reviewer: There are four target columns but six values provided.

```sql
insert into experiments (name, start_date, end_date, user_id) 

> values ('MAC Season 6: Sorghum BAP', '2018-04-06', '2018-08-01', 'some text', 'some text', 6000000004)

Reviewer: There are four target columns but six values provided.

```

### Step 2: Add new sites

> _must provide shapefile for a site to get geometry_

Reviewer: You should either provide a full explanation (in complete sentences!) or leave this out.  And actually, thinking less "TERRA-REF-centrically", you can add simple point geometries rather easily, for example
```sql
insert into sites (city, state, country, sitename, geometry) 
values ('Urbana', 'Illinois', 'United States', 'My garden plot', ST_SetSRID(ST_makePoint(88, 40, 222), 4326));
```

```sql
insert into sites (city, state, country, sitename) 
values ('Maricopa', 'Arizona', 'USA', 'MAC Field Scanner Season 7 Range 9 Column 15')
```

For the TERRA REF Project, plot definitions may be copied from previous season if same plots are used.

```sql
with season6 as (
  select city, state, replace(sitename, 'Season 4', 'Season 6') as sitename, greenhouse, geometry, time_zone from sites where sitename like '%Season 4%' 
)
insert into sites (city, state, sitename, greenhouse, geometry, time_zone) select * from season6
```

### Step 3: Add new treatments

```sql
insert into treatments (name, definition, control) 
values ('MAC Season 6: Sorghum', 'some text', 't')
```

### Step 4: Add new cultivars

_each cultivar should be associated with a specie_

> _new species must be added to species table before referencing_

Reviewer: Better: "Each cultivar must be associated with a species.  If there is no entry for the species in the species table, it must be added before adding the new cultivar."

```sql
insert into cultivars (name, specie_id) 
values ('RIL-CS27_(TX2910/(Macia/R07007)-CS44)-CSF1-PRF2-CS27', 2588)
```

### Step 5: Add new citations

```sql
insert into citations (author, year, title) 
values ('Newcomb, Maria', 2016, 'Maricopa Agricultural Center Field Activities')
```

### Step 6: Associate experiments with sites

> 

Reviewer: Since this isn't such a simple statement, a bit of an introductory explanation is in order.  Some thing like:

> As an example, this statement could be used to associate the experiment named "MAC Season 6: Sorghum BAP" with the site named "MAC Field Scanner Season 6 Range 1 Column 1 E".

```sql
insert into experiments_sites (experiment_id, site_id) values ((select id from experiments where name = 'MAC Season 6: Sorghum BAP'),
(select id from sites where sitename = 'MAC Field Scanner Season 6 Range 1 Column 1 E'))
```

> When adding a new season for the TERRA REF project, a statement like the following can be used for associating experiments and sites since MAC Field Center sites are consistently named following the format `MAC Field Scanner Season x Range a Column b`.

Reviewer: Since only one experiment is involved here, not "experiments', I would recast this sentence to something like

> When adding a new season for the TERRA REF project, a statement like the following can be used for associating the new season's experiment [or "an experiment for the new season"] with all of the new season's sites.  For example, since MAC Field Center sites are consistently named following the format `MAC Field Scanner Season x Range a Column b`, we could use the following statement to associate the experment named "MAC Season 6: Sorghum BAP" with all Season 6 sites:

```sql
insert into experiments_sites (experiment_id, site_id) 
   select e.experiment_id, s.site_id 
        from (select id as experiment_id from experiments where name = 'MAC Season 6: Sorghum BAP')  as e 
     cross join 
         (select id as site_id from sites where sitename like 'MAC Field Scanner Season 6%') as s
```

## Step 7: Associate experiments with treatments

```sql
insert into experiments_treatments (experiment_id, treatment_id) 
values ((select id from experiments where name = 'MAC Season 6: Sorghum BAP'),
(select id from treatments where name = 'MAC Season 6: Sorghum'))
```

> When adding a new season for the TERRA REF project, a statement like the following can be used to associate experiments with treatments since experiment and treatment names are usually named following the format `MAC Season x: subexperiment name`.

Reviewer: Again, this could be clearer.  At a minimum, you could say "... can be used to associate all season 6 experiments with all season 6 treatments ..." and emend the format to `MAC Season x: [sub-experiment or treatment name]` or just `MAC Season x: ...`.

```sql
insert into experiments_treatments (experiment_id, treatment_id) 
   select e.experiment_id, t.treatment_id 
        from (select id as experiment_id from experiments where name like 'MAC Season 6:%')  as e 
     cross join 
         (select id as treatment_id from treatments where name like 'MAC Season 6:%') as s
```

### Step 8: Associate sites with cultivars

> 

Reviewer: Again, a bit of explanation is in order.  And as long as you are avoiding hard-coded id numbers for sites and cultivars, you might as well do so for the species as well.

"As an example, this statement could be used to associate the site name _MAC Field Scanner Season 8 Range 1 Column 1 E_ with the Sorghum bicolor cultivar _Tiburon_:
```sql
insert into sites_cultivars (site_id, cultivar_id) 
    values ((select id from sites where sitename = 'MAC Field Scanner Season 8 Range 1 Column 1 E'),
            (select id from cultivars where name = 'Tiburon' and specie_id = (select id from species where scientificname = 'Sorghum bicolor')))
```

```sql
insert into sites_cultivars (site_id, cultivar_id) 
values ((select id from sites where sitename = 'MAC Field Scanner Season 8 Range 1 Column 1 E'),
(select id from cultivars where name = 'Tiburon' and specie_id = 2588))
```

### Step 9: Associate sites with citations

```sql
insert into citations_sites (citation_id, site_id)
values ((select id from citations where author = 'Newcomb, Maria' and year = 2016 and title = 'MAC Field Activities'),
(select id from sites where sitename = 'MAC Field Scanner Season 6 Range 1 Column 1 E'))
```

> When adding a new season for the TERRA REF project, a statement like the following can be used to associate citations with sites since MAC Field Center sites are consistently named following the format `MAC Field Scanner Season x Range a Column b`

Reviewer: Clarify more specifically what this statement is doing: "... can be used to associate all citations with author "Newcomb, Maria" with all Season 6 sites ...".

```sql
insert into citations_sites (citation_id, site_id) 
   select c.id, s.id 
      from (select id from citations where author = 'Newcomb, Maria') 
     AS c 
      cross join 
         (select id from sites where sitename like 'MAC Field Scanner Season 6%') AS s;
```

## Removing duplicate records

> 

Reviewer: If you envision people actually _using_ this function, I could write a much better version which would at least get rid of the easy-to-fix defects.

That said, I think the introduction could be much clearer about describing the use case for the function:

---
"It sometimes happens that multiple rows in a BETYdb table contain essentially the same information.  This makes it harder to obtain consistent and comprehensible search results.  A function such as the one that follows could be used to delete a duplicate row and change all references to it to point to a row that we are retaining.  (Note that there are serious problems associated with the use of this function, as outlined in `[this comment in issue 185](https://github.com/PecanProject/bety/issues/185#issuecomment-530554650){target="_blank"}`.  We present it here only to outline what might be possible along the lines of partially automating the correction of data errors.)

"The function takes three arguments: the name of the table we wish to update (as a string), the id number of the row we wish to remove (the "duplicate"), and the id number of the similar row we wish to retain.  For example, if we have two citation rows having essentially the same information having id numbers 286 and 289, we could remove the first and update references to it to point to the second with the statement
```sql
SELECT update_refs_from_to('citations', 286, 289);
```
---

The difficult issues associated with this function are all associated with the distributed data system we have come up with.  It boils down to this: If you use the function to delete a row, one of a number of things may happen:
1. If the row you deleted is owned by your database but the referring rows are owned by a different database, those references will point to nothing as soon as someone re-imports your database.  And even though you updated your copy of those referring rows using the function, the references you updated will be reverted back to point to nothing if and when you re-import from the database where they originated.
2. If both the deleted row and the referring rows are owned by your database, nothing bad should happen.  The function should update the references to point to a retained row and these updates will propagate to other databases that import yours.
3. If the row you deleted is owned by another database, it will reappear as soon as you re-import that other database.

Probably the "safest" use case for the function is this: You create a species row, say, that duplicates an existing species row on another machine.  (If that other machine is one you import from, assume perhaps that it was created since you last imported from that machine.  Equally likely is that you _have_ imported from that other machine recently but that there are no uniqueness constraints in place that prevent you from creating a row with essentially the same information as the one you imported.)  Then you add a bunch of data that refers to the new species row you created.  Then you realize your new species row is redundant and use the function to delete it and update references to it to point to the row you imported.  As long as no one has imported your new species row to their copy of the database and started adding data that refers to it, we are essentially in case 2 above and the update-references function shouldn't cause any problem.

Reviewer 2: ```suggestion
It sometimes happens that multiple rows in a BETYdb table are duplicates, e.g., two species or two sites that were independently added but reference the same entity. A function such as the one that follows could be used to delete a duplicate row and change all references to it to point to a row that we are retaining. (Note that there are serious problems associated with the use of this function, as outlined in [this comment in issue 185](https://github.com/PecanProject/bety/issues/185#issuecomment-530554650){target="_blank"}. We present it here only to outline what might be possible along the lines of partially automating the correction of data errors.)

The function takes three arguments: the name of the table we wish to update (as a string), the id number of the row we wish to remove (the "duplicate"), and the id number of the similar row we wish to retain. For example, if we have two citation rows having essentially the same information having id numbers 286 and 289, we could remove the first and update references to it to point to the second with the statement

```sql
SELECT update_refs_from_to('citations', 286, 289);
```

The issues associated with this function are associated with the distributed data system used to sync data across BETYdb instances in the PEcAn network. If you use the function to delete a row, one of a number of things may happen:

1. If the row you deleted is owned by your database but the referring rows are owned by a different database, those references will point to nothing as soon as someone re-imports your database. And even though you updated your copy of those referring rows using the function, the references you updated will be reverted back to point to nothing if and when you re-import from the database where they originated.
2. If both the deleted row and the referring rows are owned by your database, nothing bad should happen. The function should update the references to point to a retained row and these updates will propagate to other databases that import yours.
3. If the row you deleted is owned by another database, it will reappear as soon as you re-import that other database.
Probably the "safest" use case for the function is this: You create a species row, say, that duplicates an existing species row on another machine. (If that other machine is one you import from, assume perhaps that it was created since you last imported from that machine. Equally likely is that you have imported from that other machine recently but that there are no uniqueness constraints in place that prevent you from creating a row with essentially the same information as the one you imported.) Then you add a bunch of data that refers to the new species row you created. Then you realize your new species row is redundant and use the function to delete it and update references to it to point to the row you imported. As long as no one has imported your new species row to their copy of the database and started adding data that refers to it, we are essentially in case 2 above and the update-references function shouldn't cause any problem.
```

```
When adding a new season for the TERRA REF project, a statement like the following can be used to associate experiments with treatments since experiment and treatment names are usually named following the format `MAC Season x: subexperiment name`.

```sql
insert into experiments_treatments (experiment_id, treatment_id) 
   select e.experiment_id, t.treatment_id 
        from (select id as experiment_id from experiments where name like 'MAC Season 6:%')  as e 
     cross join 
         (select id as treatment_id from treatments where name like 'MAC Season 6:%') as s
```

### Step 8: Associate sites with cultivars

```sql
insert into sites_cultivars (site_id, cultivar_id) 
values ((select id from sites where sitename = 'MAC Field Scanner Season 8 Range 1 Column 1 E'),
(select id from cultivars where name = 'Tiburon' and specie_id = 2588))
```

### Step 9: Associate sites with citations

```sql
insert into citations_sites (citation_id, site_id)
values ((select id from citations where author = 'Newcomb, Maria' and year = 2016 and title = 'MAC Field Activities'),
(select id from sites where sitename = 'MAC Field Scanner Season 6 Range 1 Column 1 E'))
```

> When adding a new season for the TERRA REF project, a statement like the following can be used to associate citations with sites since MAC Field Center sites are consistently named following the format `MAC Field Scanner Season x Range a Column b`

Reviewer 2: ```suggestion

When adding a new season for the TERRA REF project, a statement like the following can be used to associate all citations with author "Newcomb, Maria" with all Season 6 sites since MAC Field Center sites are consistently named following the format `MAC Field Scanner Season x Range a Column b`

```

   select e.experiment_id, s.site_id 
        from (select id as experiment_id from experiments where name = 'MAC Season 6: Sorghum BAP')  as e 
     cross join 
         (select id as site_id from sites where sitename like 'MAC Field Scanner Season 6%') as s
```

## Step 7: Associate experiments with treatments

```sql
insert into experiments_treatments (experiment_id, treatment_id) 
values ((select id from experiments where name = 'MAC Season 6: Sorghum BAP'),
(select id from treatments where name = 'MAC Season 6: Sorghum'))
```
When adding a new season for the TERRA REF project, a statement like the following can be used to associate experiments with treatments since experiment and treatment names are usually named following the format `MAC Season x: subexperiment name`.

```sql
insert into experiments_treatments (experiment_id, treatment_id) 
   select e.experiment_id, t.treatment_id 
        from (select id as experiment_id from experiments where name like 'MAC Season 6:%')  as e 
     cross join 
         (select id as treatment_id from treatments where name like 'MAC Season 6:%') as s
```

### Step 8: Associate sites with cultivars

> 

Reviewer 2: ```suggestion

As an example, this statement could be used to associate the site name MAC Field Scanner Season 8 Range 1 Column 1 E with the Sorghum bicolor cultivar Tiburon:

insert into sites_cultivars (site_id, cultivar_id) 
    values ((select id from sites where sitename = 'MAC Field Scanner Season 8 Range 1 Column 1 E'),
            (select id from cultivars where name = 'Tiburon' and specie_id = (select id from species where scientificname = 'Sorghum bicolor')));
              
```

1. correcting systemic data errors involving large numbers of rows
2. assigning non-point geometries to sites [this might be possible in YABA, I don't know]
3. associating experiments with sites and treatments
4. associating sites with cultivars

### Metadata Entry Workflow for BETYdb

The following steps have been implemented in the BETYdb-YABA API and python client, but are described below for reference

Here we show SQL code used to add metadata required each season by the TERRA REF database, including experiments, sites, treatments, cultivars, and citations that is required prior to uploading the trait or yield data. Most PEcAn users _will not_ need to add experiments and cultivars, or associate these records with sites.

Also note that in the TERRA REF and other agronomic applications, each record in the 'sites' table will correspond to an experimental plot.

### Step 1: Add new experiments

```sql
insert into experiments (name, start_date, end_date, user_id) 
values ('MAC Season 6: Sorghum BAP', '2018-04-06', '2018-08-01', 'some text', 'some text', 6000000004);
```

### Step 2: Add new sites

> _must provide shapefile for a site to get geometry_

Reviewer 2: ```suggestion
To add sites (or many plots at a single site) using the BETYdb-YABA API, you must provide shapefile for a site to have an associated geometry. 

Otherwise, you can add sites with points or polygons to the database as follows.
```

```sql
insert into sites (city, state, country, sitename) 
values ('Maricopa', 'Arizona', 'USA', 'MAC Field Scanner Season 7 Range 9 Column 15');
```

> 

Reviewer 2: ```suggestion

You can add simple locations with a single point thus:

```sql
insert into sites (city, state, country, sitename, geometry) 
values ('Urbana', 'Illinois', 'United States', 'My garden plot', ST_SetSRID(ST_makePoint(88, 40, 222), 4326));
```
```

For the TERRA REF Project, plot definitions may be copied from previous season if same plots are used.

```sql
with season6 as (
  select city, state, replace(sitename, 'Season 4', 'Season 6') as sitename, greenhouse, geometry, time_zone from sites where sitename like '%Season 4%' 
)
insert into sites (city, state, sitename, greenhouse, geometry, time_zone) select * from season6;
```

### Step 3: Add new treatments

```sql
insert into treatments (name, definition, control) 
values ('MAC Season 6: Sorghum', 'some text', 't');
```

### Step 4: Add new cultivars

> _each cultivar should be associated with a specie_

Reviewer 2: ```suggestion
Each cultivar must be associated with a species. If there is no entry for the species in the species table, it must be added before adding the new cultivar.
```

> _new species must be added to species table before referencing_

Reviewer 2: ```suggestion
```

```sql
insert into cultivars (name, specie_id) 
values ('RIL-CS27_(TX2910/(Macia/R07007)-CS44)-CSF1-PRF2-CS27', 2588);
```

### Step 5: Add new citations

```sql
insert into citations (author, year, title) 
values ('Newcomb, Maria', 2016, 'Maricopa Agricultural Center Field Activities');
```

### Step 6: Associate experiments with sites

> 

Reviewer 2: ```suggestion

As an example, this statement could be used to associate the experiment named "MAC Season 6: Sorghum BAP" with the site named "MAC Field Scanner Season 6 Range 1 Column 1 E".
```

```sql
insert into experiments_sites (experiment_id, site_id) values 
  ((select id from experiments where name = 'MAC Season 6: Sorghum BAP'),
  (select id from sites where sitename = 'MAC Field Scanner Season 6 Range 1 Column 1 E'));
```

> When adding a new season for the TERRA REF project, a statement like the following can be used for associating experiments and sites since MAC Field Center sites are consistently named following the format `MAC Field Scanner Season x Range a Column b`.

Reviewer 2: ```suggestion
When adding a new season for the TERRA REF project, a statement like the following can be used for associating the new season's experiment [or "an experiment for the new season"] with all of the new season's sites. For example, since MAC Field Center sites are consistently named following the format MAC Field Scanner Season x Range a Column b, we could use the following statement to associate the experment named "MAC Season 6: Sorghum BAP" with all Season 6 sites:
```

```sql
insert into experiments_sites (experiment_id, site_id) 
   select e.experiment_id, s.site_id 
        from (select id as experiment_id from experiments where name = 'MAC Season 6: Sorghum BAP')  as e 
     cross join 
         (select id as site_id from sites where sitename like 'MAC Field Scanner Season 6%') as s;
```

> ## Step 7: Associate experiments with treatments

Reviewer 2: ```suggestion
### Step 7: Associate experiments with treatments
```

```sql
insert into experiments_treatments (experiment_id, treatment_id) 
values ((select id from experiments where name = 'MAC Season 6: Sorghum BAP'),
(select id from treatments where name = 'MAC Season 6: Sorghum'));
```

> When adding a new season for the TERRA REF project, a statement like the following can be used to associate experiments with treatments since experiment and treatment names are usually named following the format `MAC Season x: subexperiment name`.

Reviewer 2: ```suggestion
When adding a new season for the TERRA REF project, a statement like the following can be used to associate all season 6 experiments with all season 6 treatments assuming the experiment and treatment names follow the format convention `MAC Season x: subexperiment name`.
```

     cross join 
         (select id as site_id from sites where sitename like 'MAC Field Scanner Season 6%') as s
```

## Step 7: Associate experiments with treatments

```sql
insert into experiments_treatments (experiment_id, treatment_id) 
values ((select id from experiments where name = 'MAC Season 6: Sorghum BAP'),
(select id from treatments where name = 'MAC Season 6: Sorghum'))
```
When adding a new season for the TERRA REF project, a statement like the following can be used to associate experiments with treatments since experiment and treatment names are usually named following the format `MAC Season x: subexperiment name`.

```sql
insert into experiments_treatments (experiment_id, treatment_id) 
   select e.experiment_id, t.treatment_id 
        from (select id as experiment_id from experiments where name like 'MAC Season 6:%')  as e 
     cross join 
         (select id as treatment_id from treatments where name like 'MAC Season 6:%') as s
```

### Step 8: Associate sites with cultivars

As an example, this statement could be used to associate the site name MAC Field Scanner Season 8 Range 1 Column 1 E with the Sorghum bicolor cultivar Tiburon:

> insert into sites_cultivars (site_id, cultivar_id) 

Reviewer 2: ```suggestion
```sql
insert into sites_cultivars (site_id, cultivar_id) 
```

```sql
insert into cultivars (name, specie_id) 
values ('RIL-CS27_(TX2910/(Macia/R07007)-CS44)-CSF1-PRF2-CS27', 2588)
```

### Step 5: Add new citations

```sql
insert into citations (author, year, title) 
values ('Newcomb, Maria', 2016, 'Maricopa Agricultural Center Field Activities')
```

### Step 6: Associate experiments with sites

```sql
insert into experiments_sites (experiment_id, site_id) values ((select id from experiments where name = 'MAC Season 6: Sorghum BAP'),
(select id from sites where sitename = 'MAC Field Scanner Season 6 Range 1 Column 1 E'))
```

When adding a new season for the TERRA REF project, a statement like the following can be used for associating experiments and sites since MAC Field Center sites are consistently named following the format `MAC Field Scanner Season x Range a Column b`.

```sql
insert into experiments_sites (experiment_id, site_id) 
   select e.experiment_id, s.site_id 
        from (select id as experiment_id from experiments where name = 'MAC Season 6: Sorghum BAP')  as e 
     cross join 
         (select id as site_id from sites where sitename like 'MAC Field Scanner Season 6%') as s
```

> ## Step 7: Associate experiments with treatments

Reviewer: This needs another "#" to be at the same level as all the other steps.
